cli/manage.go

```diff
@@ -132,81 +140,71 @@ func getDiscoveryOpt(c *cli.Context) map[string]string {
-	kvDiscovery, ok := discovery.(*kvdiscovery.Discovery)
-	if !ok {
-		log.Fatal("Leader election is only supported with consul, etcd and zookeeper discovery.")
-	}
-	client := kvDiscovery.Store()
-	p := path.Join(kvDiscovery.Prefix(), leaderElectionPath)
-
-	candidate := leadership.NewCandidate(client, p, addr, leaderTTL)
-	follower := leadership.NewFollower(client, p)
-	return candidate, follower
-}
-
-func setupReplication(c *cli.Context, cluster cluster.Cluster, server *api.Server, candidate *leadership.Candidate, follower *leadership.Follower, addr string, tlsConfig *tls.Config) {
-	primary := api.NewPrimary(cluster, tlsConfig, &statusHandler{cluster, candidate, follower}, c.GlobalBool("debug"), c.Bool("cors"))
+func setupReplication(c *cli.Context, cluster cluster.Cluster, client store.Store, keyPath string, server *api.Server, candidate *leadership.Candidate, addr string, tlsConfig *tls.Config) {
+	handler := &statusHandler{cluster, candidate, nil, addr}
+	primary := api.NewPrimary(cluster, tlsConfig, handler, c.GlobalBool("debug"), c.Bool("cors"))
 	replica := api.NewReplica(primary, tlsConfig)
 
 	go func() {
 		for {
-			run(cluster, candidate, server, primary, replica)
-			time.Sleep(defaultRecoverTime)
-		}
-	}()
-
-	go func() {
-		for {
-			follow(follower, replica, addr)
-			time.Sleep(defaultRecoverTime)
+			err := run(cluster, client, keyPath, handler, server, primary, replica, addr)
+			if err != nil {
+				time.Sleep(defaultRecoverTime)
+			}
 		}
 	}()
 
 	server.SetHandler(primary)
 }
 
-func run(cl cluster.Cluster, candidate *leadership.Candidate, server *api.Server, primary *mux.Router, replica *api.Replica) {
-	electedCh, errCh := candidate.RunForElection()
+func run(cl cluster.Cluster, client store.Store, keyPath string, handler *statusHandler, server *api.Server, primary *mux.Router, replica *api.Replica, addr string) error {
+	candidate := handler.candidate
+	electedCh, candidateErrCh := candidate.RunForElection()
+
+	follower := leadership.NewFollower(client, keyPath)
+	defer follower.Stop()
+	handler.follower = follower
+	leaderCh, followerErrCh := follower.FollowElection()
+
 	var watchdog *cluster.Watchdog
+	wasLeader := false
 	for {
 		select {
 		case isElected := <-electedCh:
 			if isElected {
 				log.Info("Leader Election: Cluster leadership acquired")
+				wasLeader = true
 				watchdog = cluster.NewWatchdog(cl)
 				server.SetHandler(primary)
 			} else {
 				log.Info("Leader Election: Cluster leadership lost")
 				cl.UnregisterEventHandler(watchdog)
 				server.SetHandler(replica)
+				if wasLeader {
+					return nil
+				}
 			}
 
-		case err := <-errCh:
-			log.Error(err)
-			return
-		}
-	}
-}
-
-func follow(follower *leadership.Follower, replica *api.Replica, addr string) {
-	leaderCh, errCh := follower.FollowElection()
-	for {
-		select {
 		case leader := <-leaderCh:
 			if leader == "" {
 				continue
 			}
 			if leader == addr {
 				replica.SetPrimary("")
+				server.SetHandler(primary)
```

Comment:
Essentially, the Docker leadership election library works as a two-part system: you have one channel for figuring out who the leader is, and another for trying to become the leader. We need both parts here, but the way it's currently wired up is a bit weird: we only consider ourselves the leader if we can successfully write out to the KV store, which means that if node A is the leader and no other nodes are in the election but node A can't write to the KV store for some reason (e.g. network flakiness), node A will report that it doesn't know who the leader is (the "no elected primary cluster manager" error), even though the KV store contains the correct value for the leader.

Maybe we could switch to a model where we use the follower channel to determine whether or not we're the leader, and we don't listen to the candidate's channel at all; that is, we continually try to elect ourselves as the leader, but we ignore whether or not we've successfully written to the KV store and we only look at the current value stored in the KV store, and if it's equal to our address then we consider ourselves the leader.

One failure case here is if you have a node A and a node B who are both in a leadership election. If node A is a leader and it loses its connection to the KV store, it will continue to consider itself the primary during this time, meaning that it will still schedule containers and so on. While node A is disconnected, node B might elect itself the leader, since it can still write out to the KV store. At that point, you will effectively have two primaries, since both node A and B will be responding to API requests.